subscriber.go

```diff
@@ -27,16 +25,32 @@ func newNonBlockingSubscriber(cap int, alerter alertFunc) *nonBlockingSubscriber
 		cap = 10
 	}
 
-	return &nonBlockingSubscriber{
-		ch:    make(chan Message, cap),
+	sub := &nonBlockingSubscriber{
+		ch:    make(chan Message, 1),
+		close: make(chan struct{}, 1),
+		input: make(chan Message, cap),
 		alert: alerter,
 	}
+
+	go func() {
```

Comment:
The implementation is not wrong but it adds more complexity just to fix a panic:
- One goroutine per subscriber (can be bad for projects using a lot of subscribers)
- A copy of messages from `input` to `ch` 
- A close channel with the possibility of getting blocked (calling the Close function more than once)

The root problem is the hub allowing to unsubscribe at the same time as publishing. But this was a simple design and I need to think if there is a better way.

To fix the bug without adding this complexity I have two options:
1º - Accept that we can set on a closed channel and rescue:
```go
// Set inserts the given Event into the subscriber.
func (s *nonBlockingSubscriber) Set(msg Message) {
	defer func() {
		if recover() != nil {
			// trying to send a message on a closed subscriber, just ignore
		}
	}()
	select {
	case s.ch <- msg:
	default:
		s.alert(1)
	}
}
``` 

2º - (Maybe) Add an atomic flag for the close state and validate before sending it to the channel.

WDYT of changing the solution to one of these options?

Reply:
I agree, the first solution would be much simpler and more efficient. Changed the code to use recover